app/stores/ada/TrezorConnectStore.js

```diff
@@ -169,54 +156,125 @@ export default class TrezorConnectStore extends Store {
-        Logger.error(`TrezorConnectStore::_checkAndStoreTrezorDeviceInfo ${stringifyError(this.error)}`);
+  _normalizeHWResponse = (
+    trezorResp: any,
+    trezorEventDevice: DeviceMessage
+  ): HWDeviceInfo => {
+    this._validateHWResponse(trezorResp, trezorEventDevice);
+
+    /** This check aready done in _validateHWResponse but flow needs this */
+    if (trezorEventDevice == null
+      || trezorEventDevice.payload == null
+      || trezorEventDevice.payload.features == null) {
+      throw new UnexpectedError();
+    }
+
+    const deviceFeatures = trezorEventDevice.payload.features;
+    return {
+      publicMasterKey: trezorResp.payload.publicKey,
+      hwFeatures: {
+        vendor: deviceFeatures.vendor,
+        model: deviceFeatures.model,
+        deviceId: deviceFeatures.device_id,
+        label: deviceFeatures.label,
+        majorVersion: deviceFeatures.major_version,
+        minorVersion: deviceFeatures.minor_version,
+        patchVersion: deviceFeatures.patch_version,
+        language: deviceFeatures.language,
       }
+    };
+  }
+
+  /** Validates the compatibility of data which we have received from Trezor device */
+  _validateHWResponse = (
+    trezorResp: any,
+    trezorEventDevice: DeviceMessage
+  ): boolean => {
+
+    if (trezorResp && !trezorResp.success) {
+      switch (trezorResp.payload.error) {
+        case 'Iframe timeout':
+          throw new LocalizableError(globalMessages.trezorError101);
+        case 'Permissions not granted':
+          throw new LocalizableError(globalMessages.trezorError102);
+        case 'Cancelled':
+        case 'Popup closed':
+          throw new LocalizableError(globalMessages.trezorError103);
+        default:
+          // Something unexpected happened
+          Logger.error(`TrezorConnectStore::_validateHWResponse::error: ${trezorResp.payload.error}`);
+          throw new UnexpectedError();
+      }
+    }
+
+    if (trezorResp == null
+      || trezorResp.payload == null
+      || trezorResp.payload.publicKey == null
+      || trezorResp.payload.publicKey.length <= 0) {
+      // Something unexpected happened
+      Logger.error(`TrezorConnectStore::_validateHWResponse::error: invalid public key`);
+      throw new UnexpectedError();
+    }
+
+    if (trezorEventDevice == null
+      || trezorEventDevice.payload == null
+      || trezorEventDevice.payload.type !== 'acquired'
+      || trezorEventDevice.payload.features == null) {
+      // Something unexpected happened
+      Logger.error(`TrezorConnectStore::_validateHWResponse::error: invalid device event`);
+      throw new UnexpectedError();
     }
+
+    return true;
+  };
+
+  _handleConnectError = (error): void => {
+    Logger.error(`TrezorConnectStore::_handleConnectError ${stringifyError(error)}`);
+
+    this.hwDeviceInfo = undefined;
+
+    if (error instanceof LocalizableError) {
+      this.error = error;
+    } else {
+      this.error = new UnexpectedError();
```

Comment:
that is not swallow, it has fixed message inside it.
UnexpectedError is a LocalizableError for UI display (in case we are not sure which error happened)